lib/Lower/Bridge.cpp

```diff
@@ -574,51 +385,338 @@ class FirConverter : public Fortran::lower::AbstractConverter {
+      genFIRConditionalBranch(*whileCondition, doStmtEval.lexicalSuccessor,
+                              doStmtEval.parentConstruct->constructExit);
+    } else if (const auto *bounds{
+                   std::get_if<Fortran::parser::LoopControl::Bounds>(
+                       &loopControl->u)}) {
+      // "Normal" increment loop.
+      incrementLoopInfo.emplace_back(IncrementLoopInfo(
+          bounds->name.thing.symbol, bounds->lower, bounds->upper, bounds->step,
+          genType(*bounds->name.thing.symbol), builder->getInsertionBlock()));
+      if (unstructuredContext) {
+        maybeStartBlock(doStmtEval.block); // preheader block
+        incrementLoopInfo[0].headerBlock = doStmtEval.localBlocks[0];
+        incrementLoopInfo[0].bodyBlock = doStmtEval.lexicalSuccessor->block;
+        incrementLoopInfo[0].successorBlock =
+            doStmtEval.parentConstruct->constructExit->block;
+      }
+    } else {
+      const auto *concurrentInfo{
+          std::get_if<Fortran::parser::LoopControl::Concurrent>(
+              &loopControl->u)};
+      assert(concurrentInfo && "DO loop variant is invalid");
+      TODO();
+      // Add entries to incrementLoopInfo.  (Define extra members for a mask.)
+    }
+    for (size_t i{0}, n{incrementLoopInfo.size()}; i < n; ++i) {
+      genFIRIncrementLoopBegin(incrementLoopInfo[i]);
+    }
+
+    // Generate loop body code.
+    for (auto &e : *eval.evaluationList) {
+      genFIR(e, unstructuredContext);
+    }
 
-  void genFIR(const Fortran::parser::AssociateConstruct &) { TODO(); }
-  void genFIR(const Fortran::parser::BlockConstruct &) { TODO(); }
-  void genFIR(const Fortran::parser::ChangeTeamConstruct &) { TODO(); }
-  void genFIR(const Fortran::parser::CriticalConstruct &) { TODO(); }
-  void genFIR(const Fortran::parser::DoConstruct &) { TODO(); }
-  void genFIR(const Fortran::parser::IfConstruct &) { TODO(); }
+    // Generate end loop code.
+    if (infiniteLoop || whileCondition) {
+      genFIRUnconditionalBranch(doStmtEval.localBlocks[0]);
+    } else {
+      for (size_t i{incrementLoopInfo.size()}; i > 0;) {
+        genFIRIncrementLoopEnd(incrementLoopInfo[--i]);
+      }
+    }
+  }
 
-  void genFIR(const SelectCaseConstruct &) { TODO(); }
-  void genFIR(const SelectRankConstruct &) { TODO(); }
-  void genFIR(const SelectTypeConstruct &) { TODO(); }
+  /// Generate FIR to begin a structured or unstructured increment loop.
+  void genFIRIncrementLoopBegin(IncrementLoopInfo &info) {
+    mlir::Location location{toLocation()};
+    mlir::Type type{info.isStructured()
+                        ? mlir::IndexType::get(builder->getContext())
+                        : info.loopVariableType};
+    mlir::Value lowerValue{genFIRLoopIndex(info.lowerExpr, type)};
+    mlir::Value upperValue{genFIRLoopIndex(info.upperExpr, type)};
+    if (info.stepExpr.has_value()) {
+      info.stepValue = genFIRLoopIndex(*info.stepExpr, type);
+    }
```

Comment:
The given style URL is for early exits, rather than single statement braces, and a quick check doesn't find a braces preference.  Given my druthers, I would probably skip single statement braces, and probably put short single statements directly on the same line as the IF in at least some cases.  But here I can live with whatever.  I just need to know what that is.  And note that this can get complicated.  What about an if/else if/else sequence with all single statements?  With mixed single/multiple statements?

Reply:
Ah, another viewing found a "and better yet .." comment tucked away at the end of this style section.  That somewhat offhand comment makes it seem like an allowed pattern, rather than a firmly held requirement.  So again, although I do have some personal preferences here, I can live with any choice, as long as I know what it is.  And then it would be a good idea to apply it consistently everywhere, ideally via clang-format.  I would rather not change just one or two cases without changing others.

Reply:
From experience, I know *some* LLVM reviewers prefer the needless braces be removed. (I couldn't find a definitive mandate on the point. I just know I've been asked to delete them in the past.)

Reply:
I think I'll kick this can down the road.